File: 295-find-median-from-data-stream/295-find-median-from-data-stream.cpp
class MedianFinder {
    priority_queue<int> left, right;
public:
    MedianFinder() {
        // left.push(INT_MIN);
        // right.push(INT_MAX);
    }
    
    void addNum(int num) {
        left.push(num);
        
        right.push(-left.top());
        left.pop();
        
        if (left.size() < right.size()) {
            left.push(-right.top());
            right.pop();
        }
        // cout << left.size() << " " << right.size() << endl;
    }
    
    double findMedian() {
        if (left.size() == right.size()) {
            return ((double)(left.top() - right.top())) / 2;
        } else {
            return left.top();
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 295-find-median-from-data-stream/295-find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.minHeap = list()
        self.maxHeap = list()

    def addNum(self, num: int) -> None:
        heapq.heappush(self.minHeap, num)
        minNum = heapq.heappop(self.minHeap)
        heapq.heappush(self.maxHeap, -minNum)
        if len(self.maxHeap) - len(self.minHeap) == 2:
            minNum = heapq.heappop(self.maxHeap)
            heapq.heappush(self.minHeap, -minNum)

    def findMedian(self) -> float:
        if len(self.maxHeap) > len(self.minHeap):
            return -self.maxHeap[0]
        return (self.minHeap[0] - self.maxHeap[0]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()